pandas/core/reshape/merge.py

```diff
@@ -2407,13 +2408,20 @@ def _factorize_keys(
                 or is_string_dtype(lk.dtype)
                 and not sort
             )
+            or is_string_dtype(lk.dtype)
+            and lk.dtype.storage == "pyarrow"
```

Comment:
nitpick can you put parens around the these so i dont have to think about or/and ordering

Reply:
Added for both